File: filethumbnails/js/js/scenario6.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario6.html", {
        ready: function (element, options) {
            document.getElementById("picture-thumb-start").addEventListener("click", pickPhoto, false);
        }
    });

    function pickPhoto() {
        // Get scenario options from HTML
        var isFastSelected = document.getElementById("picture-thumb-fastThumbnail").checked;

        var requestedSize = parseInt(document.getElementById("picture-thumb-requestSize").value);
        if (requestedSize > 0 && !isNaN(requestedSize)) {
            // Set API parameters
            var thumbnailModeName = "ThumbnailMode.singleItem",
                thumbnailMode = Windows.Storage.FileProperties.ThumbnailMode.singleItem,
                thumbnailOptions = Windows.Storage.FileProperties.ThumbnailOptions.useCurrentScale;

            if (isFastSelected) {
                thumbnailOptions |= Windows.Storage.FileProperties.ThumbnailOptions.returnOnlyIfCached;
            }

            // Clean output in case of repeat usage
            cleanOutput();

            var openpicker = new Windows.Storage.Pickers.FileOpenPicker();
            openpicker.fileTypeFilter.replaceAll([".jpg", ".png", ".bmp", ".gif", ".tif"]);
            openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.picturesLibrary;
            openpicker.pickSingleFileAsync().done(function (file) {
                if (file) {
                    file.getScaledImageAsThumbnailAsync(thumbnailMode, requestedSize, thumbnailOptions).done(function (thumbnail) {
                        if (thumbnail) {
                            outputResult(file, thumbnail, thumbnailModeName, requestedSize);
                        } else if (isFastSelected) {
                            WinJS.log && WinJS.log(SdkSample.errors.noExif, "sample", "status");
                        } else {
                            WinJS.log && WinJS.log(SdkSample.errors.noThumbnail, "sample", "status");
                        }
                    }, function (error) {
                        WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                    });
                } else {
                    WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
                }
            });
        } else {
            WinJS.log && WinJS.log(SdkSample.errors.invalidSize, "sample", "error");
        }
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("picture-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("picture-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("picture-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("picture-thumb-fileName").innerText = "File used: " + item.name;
        document.getElementById("picture-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("picture-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("picture-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("picture-thumb-modeName").innerText = "";
        document.getElementById("picture-thumb-fileName").innerText = "";
        document.getElementById("picture-thumb-requestedSize").innerText = "";
        document.getElementById("picture-thumb-returnedSize").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario6.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario6.h"
#include "Scenario6.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace
{
    // Utility function to convert a string to a uint32_t and detect bad input
    bool TryParseUInt(const wchar_t* str, uint32_t& result)
    {
        wchar_t* end;
        errno = 0;
        result = std::wcstoul(str, &end, 0);

        if (str == end)
        {
            // Not parseable.
            return false;
        }

        if (errno == ERANGE)
        {
            // Out of range.
            return false;
        }

        if (*end != L'\0')
        {
            // Extra unparseable characters at the end.
            return false;
        }

        return true;
    }

}

namespace winrt::SDKTemplate::implementation
{
    Scenario6::Scenario6()
    {
        InitializeComponent();
    }

    void Scenario6::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());
    }

    fire_and_forget Scenario6::SelectImageButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());

        uint32_t size;
        if (TryParseUInt(RequestSize().Text().c_str(), size))
        {
            // Pick a photo
            FileOpenPicker openPicker;
            openPicker.FileTypeFilter().ReplaceAll(FileExtensions::Image());
            StorageFile file = co_await openPicker.PickSingleFileAsync();
            if (file != nullptr)
            {
                const ThumbnailMode thumbnailMode = ThumbnailMode::SingleItem;

                bool fastThumbnail = FastThumbnailCheckBox().IsChecked().Value();
                ThumbnailOptions thumbnailOptions = ThumbnailOptions::UseCurrentScale;
                if (fastThumbnail)
                {
                    thumbnailOptions |= ThumbnailOptions::ReturnOnlyIfCached;
                }

                StorageItemThumbnail thumbnail = co_await file.GetScaledImageAsThumbnailAsync(thumbnailMode, size, thumbnailOptions);
                    if (thumbnail != nullptr)
                    {
                        Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, file, thumbnail, false);
                        thumbnail.Close();
                    }
                    else if (fastThumbnail)
                    {
                        rootPage.NotifyUser(Errors::NoExifThumbnail, NotifyType::StatusMessage);
                    }
                    else
                    {
                        rootPage.NotifyUser(Errors::NoThumbnail, NotifyType::StatusMessage);
                    }
            }
            else
            {
                rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::InvalidSize, NotifyType::ErrorMessage);
        }
    }
}
